### PR TITLE
Spearman and Pearson plots are now made and saved in the results directory; Added matplotlib to requirements.txt

### LINKED ISSUE #37: Plot predictions vs targets and include scores

### PR DESCRIPTION
**Pull request by erikhuck**
closes #37

### CODE REVIEW

File: dna/metrics.py [hunk 1]
```diff
@@ -15,11 +17,22 @@ def rmse(y_hat, y):
     return mean_squared_error(y, y_hat)**.5
 
 
-def pearson_correlation(y_hat, y):
+def pearson_correlation(y_hat, y, name: str, directory: str):
```
Comment (bjschoenfeld) [inline dna/metrics.py]: Let's keep functions as simple as possible. Making the plots should be separate from the score computation. Also, the score should be included in the plot, as I linked here: https://github.com/byu-dml/d3m-dynamic-neural-architecture/issues/37#issuecomment-501427904.

Comment (erikhuck) [reply dna/metrics.py]: Done

Comment (bjschoenfeld) [reply dna/metrics.py]: Use the "Resolve Conversation" button to let me know that it is resolved.

File: dna/metrics.py [hunk 2]
```diff
@@ -45,5 +45,8 @@ def top_k_regret(ranked_data: pd.DataFrame, actual_data: pd.DataFrame, k: int):
 
 def spearman_correlation(ranked_data: pd.DataFrame, actual_data: pd.DataFrame):
     actual_data = pd.DataFrame(actual_data)
-    score = scipy.stats.spearmanr(ranked_data['rank'], utils.rank(actual_data.test_f1_macro))
+    ranked_data = ranked_data['rank']
+    actual_data = utils.rank(actual_data.test_f1_macro)
+    score = scipy.stats.spearmanr(ranked_data, actual_data)
```
Comment (bjschoenfeld) [inline dna/metrics.py]: Try to keep the changes in a single PR scoped as much as possible. This change is not related to the topic of this PR, so it makes it harder to review and potentially longer before it gets merged. For example, as the reviewer, if I approved of everything in the PR except this unrelated changed, I would not merge the PR.

File: dna/problems.py [hunk 3]
```diff
@@ -34,13 +35,13 @@ def run(
         train_predictions = model_predict_method(train_data, verbose=verbose, **predict_regression_model_config)
         test_predictions = model_predict_method(test_data, verbose=verbose, **predict_regression_model_config)
 
-        train_scores = self._score(train_predictions, train_data)
-        test_scores = self._score(test_predictions, test_data)
+        train_scores = self._score(train_predictions, train_data, graph_name='train', graph_directory=output_dir)
+        test_scores = self._score(test_predictions, test_data, graph_name='validation', graph_directory=output_dir)
```
Comment (bjschoenfeld) [inline dna/problems.py]: As before, let's keep scoring and plotting code separate.

File: dna/problems.py [hunk 4]
```diff
@@ -90,8 +93,8 @@ def run(
         train_predicted_ranks = self._predict_rank(train_data_by_dataset, model, verbose, predict_rank_model_config)
         test_predicted_ranks = self._predict_rank(test_data_by_dataset, model, verbose, predict_rank_model_config)
 
-        train_scores = self._score(scores, train_predicted_ranks, train_data_by_dataset, k)
-        test_scores = self._score(scores, test_predicted_ranks, test_data_by_dataset, k)
+        train_scores = self._score(scores, train_predicted_ranks, train_data_by_dataset, k, output_dir)
```
Comment (bjschoenfeld) [inline dna/problems.py]: Separate scoring and plotting code.

File: dna/utils.py [hunk 5]
```diff
@@ -1,4 +1,37 @@
 import numpy as np
+import pandas as pd
+from matplotlib import pyplot as plt
+import os
 
 def rank(values):
     return np.argsort(values)[::-1]
+
+
+def plot_spearman(ranked_data: pd.DataFrame, actual_data: pd.DataFrame, name: str, directory: str, score: float):
```
Comment (bjschoenfeld) [inline dna/utils.py]: Let's put as many scores in on the same plot as possible, similar to how it is in the [link](https://github.com/byu-dml/d3m-dynamic-neural-architecture/issues/37#issuecomment-501427904), instead of having one plot per score. This allows us to compare all the scores together.

File: dna/metrics.py (hunk 2 above)
Comment (erikhuck) [reply dna/metrics.py]: This change is a result of incorporating plotting into the score function and then separating it from the score function. At this point, I don't remember exactly what the function definition was before but I can copy it from a previous commit.

File: dna/utils.py (hunk 5 above)
Comment (erikhuck) [reply dna/utils.py]: Are you saying you want to combine spearman and pearson scores into the same plot?

Comment (bjschoenfeld) [reply dna/utils.py]: Yes, I want them in the same plot. All scores are computed using the same predictions and targets. Make the code dynamic so that it includes all the scores that are provided.

File: dna/metrics.py (hunk 2 above)
Comment (bjschoenfeld) [reply dna/metrics.py]: Yes, just look at the git history.

**Comment by erikhuck** [comment]
All the plots work except for autosklearn. For some reason, the length of its predicted values do not match the length of its actual values.

**Comment by bjschoenfeld** [comment]
This is a great start. A few suggestions:

The plots should not go under the `model` directory. That is reserved for the model itself to do what it needs.

The names of the plots are are being stacked in a weird way. There is probably an internal state variable that is persisting between plots. For example:

train_LL0_1100_popularkids_dataset_185_baseball_dataset.png
and
train_LL0_1100_popularkids_dataset_185_baseball_dataset_LL0_13_breast_cancer_dataset.png

Make sure none of the points in the plot are on the boundary lines of the plot.

Some of the axes have too big of a range.

**Comment by bjschoenfeld** [comment]
I also got an error:

```
ValueError: Format '2_50_edm_2_001_dataset' is not supported (supported formats: eps, pdf, pgf, png, ps, raw, rgba, svg, svgz)
```

Its probably related to the naming issue.

**Comment by bjschoenfeld** [comment]
@erikhuck Didn't we already merge in develop? Did you not push it?

**Comment by erikhuck** [comment]
My mistake. It's been pushed now.
